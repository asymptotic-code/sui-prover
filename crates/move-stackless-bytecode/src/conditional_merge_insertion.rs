// Copyright (c) Asymptotic
// SPDX-License-Identifier: Apache-2.0

//! Inserts `if_then_else` at merge points for simple consecutive if-statements
//! while keeping original control flow intact. This is intended to make merge
//! semantics explicit for later verification/optimization passes.
//!
//! Supported pattern per if-block:
//! - Bytecode::Branch(then_label, else_label, cond)
//! - then_label block ends with Jump to else_label
//! - else_label is the merge label (fallthrough path)
//! - Simplifying assumption: pick the last Assign in the then block as
//!   the then-value producer for the merged variable, and the last assignment
//!   to the same variable before the Branch as the else-value.
//!
//! The pass inserts at the merge label (right after the Label):
//!   t_new := if_then_else(cond, then_src, else_src)
//!   var   := t_new
//!
//! Branch-local assignments are left in place (they may be dead after merge
//! and can be cleaned up by later passes).

use crate::{
    exp_generator::ExpGenerator,
    function_data_builder::FunctionDataBuilder,
    function_target::FunctionData,
    function_target_pipeline::{FunctionTargetProcessor, FunctionTargetsHolder},
    livevar_analysis::LiveVarAnnotation,
    stackless_bytecode::{AssignKind, AttrId, Bytecode, Label, Operation},
    stackless_control_flow_graph::StacklessControlFlowGraph,
};
use move_compiler::shared::known_attributes::AttributeKind_;
use move_model::model::FunctionEnv;
use std::collections::{BTreeMap, BTreeSet};
use std::mem;

pub struct ConditionalMergeInsertionProcessor {
    debug: bool,
}

impl ConditionalMergeInsertionProcessor {
    pub fn new() -> Box<Self> {
        Box::new(Self { debug: false })
    }

    #[allow(dead_code)]
    pub fn new_with_debug() -> Box<Self> {
        Box::new(Self { debug: true })
    }

    // Dominator-based merge detection for a branch at pc.
    // Returns the merge label if we can determine a valid merge pattern.
    //
    // Recall the immediate postdominator of the branch corresponds to the
    // structural join. For SSA-style placement of `if_then_else` (phi) per
    // variable, we compute the dominance frontier of each definition and place
    // merges at those DF nodes. This function currently finds the structural
    // join (later phases can refine per-var merge placement using DF).
    fn find_merge_by_dominators(
        &self,
        code: &[Bytecode],
        back_cfg: &StacklessControlFlowGraph,
        branch_pc: usize,
    ) -> Option<Label> {
        let labels = Bytecode::label_offsets(code);
        let branch_block = StacklessControlFlowGraph::pc_to_block(back_cfg, branch_pc as u16)?;
        // Compute merge block as immediate post-dominator of the branch block using
        // reversed-graph dominator analysis (postdominators of the original graph).
        let merge_block =
            StacklessControlFlowGraph::find_immediate_post_dominator(back_cfg, branch_block)?;
        let merge_pc = StacklessControlFlowGraph::block_start_pc(back_cfg, merge_block)?;

        // Find the label that corresponds to merge_pc
        labels
            .iter()
            .find(|(_, &pc)| pc == merge_pc)
            .map(|(label, _)| *label)
    }

    // Helper function to scan a block for assignments before any nested control flow
    // For nested conditionals, only considers assignments that happen directly in this block
    // Returns all assigned variables and their sources
    fn block_permitted(
        &self,
        orig_code: &[Bytecode],
        start_pc: usize,
        end_pc: usize,
    ) -> Option<BTreeMap<usize, usize>> {
        let mut scan_pc = start_pc;
        let mut last_assign_per_var: BTreeMap<usize, usize> = BTreeMap::new();
        let mut found_nested_control = false;

        while scan_pc < orig_code.len() && scan_pc < end_pc {
            match &orig_code[scan_pc] {
                Bytecode::Assign(_, dst, src, _) => {
                    if !found_nested_control {
                        // Only record assignments before nested control flow
                        last_assign_per_var.insert(*dst, *src);
                    }
                }
                Bytecode::Call(_, dests, _op, _srcs, _aa) if !dests.is_empty() => {
                    if !found_nested_control {
                        // Only record calls before nested control flow
                        let var = dests[0];
                        last_assign_per_var.insert(var, var);
                    }
                }
                Bytecode::Branch(..) => {
                    // Mark that we found nested control flow - stop considering new assignments
                    found_nested_control = true;
                }
                Bytecode::Label(..) => {
                    if scan_pc != start_pc {
                        // We've entered a nested block, stop considering assignments
                        found_nested_control = true;
                    }
                }
                Bytecode::Jump(..) => {
                    // Expected: blocks should jump to merge point
                }
                Bytecode::Load(..) => {}
                _ => {
                    // Other instructions don't affect our analysis
                }
            }
            scan_pc += 1;
        }

        if !last_assign_per_var.is_empty() {
            Some(last_assign_per_var)
        } else {
            None
        }
    }

    // Build conditional insert using postdominator analysis results
    // Handles both if-then (fallthrough) and if-then-else patterns uniformly
    // Now returns multiple insertions for live assigned variables
    fn build_conditional_insert(
        &self,
        code: &[Bytecode],
        then_label: Label,
        else_label: Label,
        merge_label: Label,
        cond_idx: usize,
        src_attr: AttrId,
        current_val: &BTreeMap<usize, usize>,
        builder: &mut FunctionDataBuilder,
        liveness: Option<&LiveVarAnnotation>,
    ) -> Option<(Label, Vec<Insertion>)> {
        let labels = Bytecode::label_offsets(code);
        let then_pc = *labels.get(&then_label)? as usize;
        let else_pc = *labels.get(&else_label)? as usize;
        let merge_pc = *labels.get(&merge_label)? as usize;

        // Scan the then-block: from then_pc to else_pc
        let then_assignments = self.block_permitted(code, then_pc, else_pc)?;

        // Scan the else-block: from else_pc to merge_pc
        // If else_pc == merge_pc, there's no else block (pure fallthrough)
        let else_assignments = if else_pc < merge_pc {
            self.block_permitted(code, else_pc, merge_pc)
        } else {
            None
        };

        let mut insertions = Vec::new();

        // Get live variables at the merge point to filter unnecessary insertions
        let live_vars_at_merge = if let Some(liveness) = liveness {
            liveness.get_live_var_info_at(merge_pc as u16)
                .map(|info| &info.after)
        } else {
            None
        };

        // Collect all variables assigned in either then-block or else-block
        let mut all_assigned_vars: BTreeSet<usize> = BTreeSet::new();
        all_assigned_vars.extend(then_assignments.keys());
        if let Some(else_map) = &else_assignments {
            all_assigned_vars.extend(else_map.keys());
        }

        // Create insertions for all live assigned variables
        for &var in &all_assigned_vars {
            // Skip if we have liveness info and this variable is not live at merge
            if let Some(live_vars) = live_vars_at_merge {
                if !live_vars.contains(&var) {
                    continue;
                }
            }

            let var_ty = builder.get_local_type(var);
            let then_src_opt = then_assignments.get(&var);
            let else_src_opt = else_assignments.as_ref().and_then(|m| m.get(&var));

            let mut fresh_vars_in_then = BTreeMap::new();
            let mut fresh_vars_in_else = BTreeMap::new();

            let (then_var, else_var) = match (then_src_opt, else_src_opt) {
                (Some(&then_src), Some(&else_src)) => {
                    // Both blocks assign: create fresh variables for both
                    let fresh_then_var = builder.new_temp(var_ty.clone());
                    let fresh_else_var = builder.new_temp(var_ty);
                    fresh_vars_in_then.insert(var, fresh_then_var);
                    fresh_vars_in_else.insert(var, fresh_else_var);
                    (then_src, else_src)
                }
                (Some(&then_src), None) => {
                    // Only then block assigns: fresh then var, fallthrough else var
                    let fresh_then_var = builder.new_temp(var_ty);
                    fresh_vars_in_then.insert(var, fresh_then_var);
                    let fallthrough = *current_val.get(&var).unwrap_or(&var);
                    (then_src, fallthrough)
                }
                (None, Some(&else_src)) => {
                    // Only else block assigns: fallthrough then var, fresh else var
                    let fresh_else_var = builder.new_temp(var_ty);
                    fresh_vars_in_else.insert(var, fresh_else_var);
                    let fallthrough = *current_val.get(&var).unwrap_or(&var);
                    (fallthrough, else_src)
                }
                (None, None) => {
                    // Neither block assigns (shouldn't happen due to collection logic)
                    continue;
                }
            };

            let insertion = Insertion {
                dest_var: var,
                cond: cond_idx,
                then_var,
                src_attr,
                else_var,
                fresh_vars_in_then,
                fresh_vars_in_else,
            };

            insertions.push(insertion);
        }

        if insertions.is_empty() {
            None
        } else {
            Some((merge_label, insertions))
        }
    }
}

#[derive(Clone, Debug)]
struct Insertion {
    src_attr: AttrId,
    dest_var: usize,
    cond: usize,
    then_var: usize,
    else_var: usize,
    fresh_vars_in_then: BTreeMap<usize, usize>,
    fresh_vars_in_else: BTreeMap<usize, usize>,
}

impl FunctionTargetProcessor for ConditionalMergeInsertionProcessor {
    fn process(
        &self,
        targets: &mut FunctionTargetsHolder,
        func_env: &FunctionEnv,
        data: FunctionData,
        _scc_opt: Option<&[FunctionEnv]>,
    ) -> FunctionData {
        if targets.is_spec(&func_env.get_qualified_id()) {
            return data; // Do not touch specs
        }
        if func_env
            .get_toplevel_attributes()
            .get_(&AttributeKind_::SpecOnly)
            .is_some()
        {
            return data; // Do not touch spec_only
        }

        let mut builder = FunctionDataBuilder::new(func_env, data);
        let orig_code = builder.data.code.clone();
        if orig_code.is_empty() {
            return builder.data;
        }

        // Get liveness annotation if available (clone to avoid borrowing issues)
        let liveness = builder.data.annotations.get::<LiveVarAnnotation>().cloned();

        // XXX(@rvantonder): For simplicity, rejects functions with early returns (no Ret at the end)
        let last_idx = orig_code.len().saturating_sub(1);
        let has_early_ret = orig_code
            .iter()
            .enumerate()
            .any(|(i, bc)| matches!(bc, Bytecode::Ret(..)) && i < last_idx);
        if has_early_ret {
            // Punt: do not transform this function
            return builder.data;
        }
        let mut current_val: BTreeMap<usize, usize> = (0..builder.data.local_types.len())
            .map(|i| (i, i))
            .collect();
        // pending inserts track where to insert a if_then_else(...)'s after a Label
        let mut pending_inserts: BTreeMap<Label, Vec<Insertion>> = BTreeMap::new();
        let back_cfg = Some(StacklessControlFlowGraph::new_backward(&orig_code, false));

        let mut active_fresh_vars: BTreeMap<usize, usize> = BTreeMap::new();
        let mut label_to_fresh_vars: BTreeMap<Label, BTreeMap<usize, usize>> = BTreeMap::new();

        // Find all conditionals first
        let conditionals = self.find_all_conditionals(&orig_code, back_cfg.as_ref().unwrap());


        // Process all conditionals to collect insertions
        for conditional in conditionals {
            if let Some((insertion_label, insertions)) = self.build_conditional_insert(
                &orig_code,
                conditional.then_label,
                conditional.else_label,
                conditional.merge_label,
                conditional.cond_idx,
                conditional.attr,
                &current_val,
                &mut builder,
                liveness.as_ref(),
            ) {
                // Collect fresh variables for then/else blocks from all insertions
                let mut fresh_vars_for_then_block = BTreeMap::new();
                let mut fresh_vars_for_else_block = BTreeMap::new();

                for insertion in &insertions {
                    // Collect then-block fresh vars
                    for (&var, &fresh_var) in &insertion.fresh_vars_in_then {
                        fresh_vars_for_then_block.insert(var, fresh_var);
                    }
                    // Collect else-block fresh vars
                    for (&var, &fresh_var) in &insertion.fresh_vars_in_else {
                        fresh_vars_for_else_block.insert(var, fresh_var);
                    }
                }

                if !fresh_vars_for_then_block.is_empty() {
                    label_to_fresh_vars.insert(conditional.then_label, fresh_vars_for_then_block);
                }
                if !fresh_vars_for_else_block.is_empty() {
                    label_to_fresh_vars.insert(conditional.else_label, fresh_vars_for_else_block);
                }

                pending_inserts
                    .entry(insertion_label)
                    .or_default()
                    .extend(insertions);
            }
        }

        // Now emit the transformed code
        let mut pc = 0;
        for bc in mem::take(&mut builder.data.code) {
            match bc {
                // Check labels for inserting if_then_else(...)
                Bytecode::Label(attr_id, label) => {
                    builder.emit(Bytecode::Label(attr_id, label));

                    if let Some(fresh_vars_map) = label_to_fresh_vars.get(&label) {
                        active_fresh_vars = fresh_vars_map.clone();
                    } else {
                        active_fresh_vars.clear();
                    }

                    if let Some(list) = pending_inserts.remove(&label) {
                        for ins in list {
                            // Set location/debug context from the originating branch instruction
                            builder.set_loc_from_attr(ins.src_attr);
                            // Allocate a fresh temp of same type as `var`
                            let var_ty = builder.get_local_type(ins.dest_var);
                            let new_temp = builder.new_temp(var_ty);

                            let fresh_then_var = ins
                                .fresh_vars_in_then
                                .get(&ins.dest_var)
                                .unwrap_or(&ins.then_var);

                            let fresh_else_var = ins
                                .fresh_vars_in_else
                                .get(&ins.dest_var)
                                .unwrap_or(&ins.else_var);

                            builder.set_next_debug_comment(format!(
                                "conditional_merge_insertion: t{} := if_then_else(t{}, t{}, t{})",
                                new_temp, ins.cond, fresh_then_var, fresh_else_var
                            ));
                            builder.emit_with(|id| {
                                Bytecode::Call(
                                    id,
                                    vec![new_temp],
                                    Operation::IfThenElse,
                                    vec![ins.cond, *fresh_then_var, *fresh_else_var],
                                    None,
                                )
                            });

                            builder.set_next_debug_comment(format!(
                                "conditional_merge_insertion: merge assign t{} := t{}",
                                ins.dest_var, new_temp
                            ));
                            builder.emit_with(|id| {
                                Bytecode::Assign(id, ins.dest_var, new_temp, AssignKind::Store)
                            });
                            current_val.insert(ins.dest_var, new_temp);
                        }
                    }
                }
                Bytecode::Branch(attr, then_label, else_label, cond_idx) => {
                    // Just emit the branch - insertions are handled above
                    builder.emit(Bytecode::Branch(attr, then_label, else_label, cond_idx))
                }
                Bytecode::Assign(attr, original_dst, src, kind) => {
                    let fresh_dst = active_fresh_vars
                        .get(&original_dst)
                        .copied()
                        .unwrap_or(original_dst);
                    current_val.insert(original_dst, src); // Track latest rhs value of original dst
                    builder.emit(Bytecode::Assign(attr, fresh_dst, src, kind))
                }
                Bytecode::Call(attr, dests, op, srcs, abort) if dests.len() == 1 => {
                    let original_dest = dests[0];
                    let actual_dest = active_fresh_vars
                        .get(&original_dest)
                        .copied()
                        .unwrap_or(original_dest);
                    current_val.insert(original_dest, original_dest);
                    builder.emit(Bytecode::Call(attr, vec![actual_dest], op, srcs, abort))
                }
                bytecode => builder.emit(bytecode),
            }
            pc += 1;
        }

        builder.data
    }

    fn name(&self) -> String {
        "conditional_merge_insertion".to_string()
    }
}

impl ConditionalMergeInsertionProcessor {
    // Find all conditional structures in the function
    fn find_all_conditionals(
        &self,
        code: &[Bytecode],
        back_cfg: &StacklessControlFlowGraph,
    ) -> Vec<ConditionalInfo> {
        let mut conditionals = Vec::new();

        for (pc, bytecode) in code.iter().enumerate() {
            if let Bytecode::Branch(attr, then_label, else_label, cond_idx) = bytecode {
                if let Some(merge_label) = self.find_merge_by_dominators(code, back_cfg, pc) {
                    conditionals.push(ConditionalInfo {
                        pc,
                        attr: *attr,
                        then_label: *then_label,
                        else_label: *else_label,
                        merge_label,
                        cond_idx: *cond_idx,
                    });
                }
            }
        }

        // Sort by merge label position (process inner-most first)
        let labels = Bytecode::label_offsets(code);
        conditionals.sort_by_key(|c| labels.get(&c.merge_label).unwrap_or(&0));

        conditionals
    }
}

#[derive(Clone, Debug)]
struct ConditionalInfo {
    pc: usize,
    attr: AttrId,
    then_label: Label,
    else_label: Label,
    merge_label: Label,
    cond_idx: usize,
}
