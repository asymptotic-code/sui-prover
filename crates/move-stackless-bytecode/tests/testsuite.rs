// Copyright (c) The Diem Core Contributors
// Copyright (c) The Move Contributors
// SPDX-License-Identifier: Apache-2.0

use anyhow::anyhow;
use codespan_reporting::{diagnostic::Severity, term::termcolor::Buffer};
use move_command_line_common::insta_assert;
use move_compiler::{diagnostics::warning_filters::WarningFiltersBuilder, shared::PackagePaths};
use move_model::{model::GlobalEnv, options::ModelBuilderOptions, run_model_builder_with_options};
use move_stackless_bytecode::{
    borrow_analysis::BorrowAnalysisProcessor,
    clean_and_optimize::CleanAndOptimizeProcessor,
    eliminate_imm_refs::EliminateImmRefsProcessor,
    escape_analysis::EscapeAnalysisProcessor,
    function_target_pipeline::{
        FunctionTargetPipeline, FunctionTargetsHolder, ProcessorResultDisplay,
    },
    livevar_analysis::LiveVarAnalysisProcessor,
    memory_instrumentation::MemoryInstrumentationProcessor,
    mut_ref_instrumentation::MutRefInstrumenter,
    options::ProverOptions,
    print_targets_for_test,
    reaching_def_analysis::ReachingDefProcessor,
};
use regex::Regex;
use std::{fs::File, io::Read, path::Path};

// Extracts lines out of some text file where each line starts with `start` which can be a regular
// expressions. Returns the list of such lines with `start` stripped. Use as in
// `extract_test_directives(file, "// dep:")`.
fn extract_test_directives(path: &Path, start: &str) -> anyhow::Result<Vec<String>> {
    let rex = Regex::new(&format!("(?m)^{}(?P<ann>.*?)$", start)).unwrap();
    let mut content = String::new();
    let mut file = File::open(path)?;
    file.read_to_string(&mut content)?;
    let mut at = 0;
    let mut res = vec![];
    while let Some(cap) = rex.captures(&content[at..]) {
        res.push(cap.name("ann").unwrap().as_str().trim().to_string());
        at += cap.get(0).unwrap().end();
    }
    Ok(res)
}

fn get_tested_transformation_pipeline(
    dir_name: &str,
) -> anyhow::Result<Option<FunctionTargetPipeline>> {
    match dir_name {
        "from_move" => Ok(None),
        "eliminate_imm_refs" => {
            let mut pipeline = FunctionTargetPipeline::default();
            pipeline.add_processor(EliminateImmRefsProcessor::new());
            Ok(Some(pipeline))
        }
        "mut_ref_instrumentation" => {
            let mut pipeline = FunctionTargetPipeline::default();
            pipeline.add_processor(EliminateImmRefsProcessor::new());
            pipeline.add_processor(MutRefInstrumenter::new());
            Ok(Some(pipeline))
        }
        "reaching_def" => {
            let mut pipeline = FunctionTargetPipeline::default();
            pipeline.add_processor(EliminateImmRefsProcessor::new());
            pipeline.add_processor(MutRefInstrumenter::new());
            pipeline.add_processor(ReachingDefProcessor::new());
            Ok(Some(pipeline))
        }
        "conditional_merge_insertion" => {
            use move_stackless_bytecode::conditional_merge_insertion::ConditionalMergeInsertionProcessor;
            let mut pipeline = FunctionTargetPipeline::default();
            pipeline.add_processor(EliminateImmRefsProcessor::new());
            pipeline.add_processor(MutRefInstrumenter::new());
            pipeline.add_processor(ReachingDefProcessor::new());
            pipeline.add_processor(LiveVarAnalysisProcessor::new());
            pipeline.add_processor(ConditionalMergeInsertionProcessor::new_with_debug());
            Ok(Some(pipeline))
        }
        "livevar" => {
            let mut pipeline = FunctionTargetPipeline::default();
            pipeline.add_processor(EliminateImmRefsProcessor::new());
            pipeline.add_processor(MutRefInstrumenter::new());
            pipeline.add_processor(ReachingDefProcessor::new());
            pipeline.add_processor(LiveVarAnalysisProcessor::new());
            Ok(Some(pipeline))
        }
        "borrow" => {
            let mut pipeline = FunctionTargetPipeline::default();
            pipeline.add_processor(EliminateImmRefsProcessor::new());
            pipeline.add_processor(MutRefInstrumenter::new());
            pipeline.add_processor(ReachingDefProcessor::new());
            pipeline.add_processor(LiveVarAnalysisProcessor::new());
            pipeline.add_processor(BorrowAnalysisProcessor::new());
            Ok(Some(pipeline))
        }
        "borrow_strong" => {
            let mut pipeline = FunctionTargetPipeline::default();
            pipeline.add_processor(EliminateImmRefsProcessor::new());
            pipeline.add_processor(MutRefInstrumenter::new());
            pipeline.add_processor(ReachingDefProcessor::new());
            pipeline.add_processor(LiveVarAnalysisProcessor::new());
            pipeline.add_processor(BorrowAnalysisProcessor::new());
            Ok(Some(pipeline))
        }
        "escape_analysis" => {
            let mut pipeline = FunctionTargetPipeline::default();
            pipeline.add_processor(Box::new(EscapeAnalysisProcessor {}));
            Ok(Some(pipeline))
        }
        "memory_instr" => {
            let mut pipeline = FunctionTargetPipeline::default();
            pipeline.add_processor(EliminateImmRefsProcessor::new());
            pipeline.add_processor(MutRefInstrumenter::new());
            pipeline.add_processor(ReachingDefProcessor::new());
            pipeline.add_processor(LiveVarAnalysisProcessor::new());
            pipeline.add_processor(BorrowAnalysisProcessor::new());
            pipeline.add_processor(MemoryInstrumentationProcessor::new());
            Ok(Some(pipeline))
        }
        "clean_and_optimize" => {
            let mut pipeline = FunctionTargetPipeline::default();
            pipeline.add_processor(EliminateImmRefsProcessor::new());
            pipeline.add_processor(MutRefInstrumenter::new());
            pipeline.add_processor(ReachingDefProcessor::new());
            pipeline.add_processor(LiveVarAnalysisProcessor::new());
            pipeline.add_processor(BorrowAnalysisProcessor::new());
            pipeline.add_processor(MemoryInstrumentationProcessor::new());
            pipeline.add_processor(CleanAndOptimizeProcessor::new());
            Ok(Some(pipeline))
        }
        _ => Err(anyhow!(
            "the sub-directory `{}` has no associated pipeline to test",
            dir_name
        )),
    }
}

fn test_runner(path: &Path) -> datatest_stable::Result<()> {
    // Allow opting out of external deps (e.g., move-stdlib) so isolated tests can run
    // without requiring a local stdlib checkout. Set env var `STACKLESS_TEST_IGNORE_DEPS=1`.
    let ignore_deps = std::env::var("STACKLESS_TEST_IGNORE_DEPS").is_ok();
    let mut sources = if ignore_deps {
        vec![path.to_string_lossy().to_string()]
    } else {
        let mut deps = extract_test_directives(path, "// dep:")?;
        // Keep only deps that exist on disk to avoid hard failures on missing stdlib
        deps.retain(|p| std::path::Path::new(p).exists());
        deps.push(path.to_string_lossy().to_string());
        deps
    };
    let env: GlobalEnv = run_model_builder_with_options(
        vec![PackagePaths {
            name: None,
            paths: sources,
            named_address_map: move_stdlib::named_addresses(),
        }],
        vec![],
        ModelBuilderOptions::default(),
        Some(WarningFiltersBuilder::unused_warnings_filter_for_test()),
    )?;
    let out = if env.has_errors() {
        let mut error_writer = Buffer::no_color();
        env.report_diag(&mut error_writer, Severity::Error);
        String::from_utf8_lossy(&error_writer.into_inner()).to_string()
    } else {
        let options = ProverOptions {
            stable_test_output: true,
            ..Default::default()
        };
        env.set_extension(options);
        let dir_name = path
            .parent()
            .and_then(|p| p.file_name())
            .and_then(|p| p.to_str())
            .ok_or_else(|| anyhow!("bad file name"))?;
        let pipeline_opt = get_tested_transformation_pipeline(dir_name)?;

        // Initialize and print function targets
        let mut text = String::new();
        let mut targets = FunctionTargetsHolder::new(None);
        for module_env in env.get_modules() {
            for func_env in module_env.get_functions() {
                targets.add_target(&func_env);
            }
        }
        let show_livevars = std::env::var("STACKLESS_TEST_SHOW_LIVENESS").is_ok();
        let show_borrow = true;
        let show_reach = std::env::var("STACKLESS_TEST_SHOW_REACHING_DEFS").is_ok();
        text += &move_stackless_bytecode::print_targets_for_test_with_flags(
            &env,
            "initial translation from Move",
            &targets,
            show_livevars,
            show_borrow,
            show_reach,
        );

        // Run pipeline if any
        if let Some(pipeline) = pipeline_opt {
            let _ = pipeline.run(&env, &mut targets);
            let processor = pipeline.last_processor();
            if !processor.is_single_run() {
                text += &move_stackless_bytecode::print_targets_for_test_with_flags(
                    &env,
                    &format!("after pipeline `{}`", dir_name),
                    &targets,
                    show_livevars,
                    show_borrow,
                    show_reach,
                );
            }
            text += &ProcessorResultDisplay {
                env: &env,
                targets: &targets,
                processor,
            }
            .to_string();
        }
        // add Warning and Error diagnostics to output
        let mut error_writer = Buffer::no_color();
        if env.has_errors() || env.has_warnings() {
            env.report_diag(&mut error_writer, Severity::Warning);
            text += "============ Diagnostics ================\n";
            text += &String::from_utf8_lossy(&error_writer.into_inner());
        }
        text
    };
    insta_assert! {
        input_path: path,
        contents: out,
    };
    Ok(())
}

datatest_stable::harness! {
    { test = test_runner, root = "tests", pattern = r".*\.move" },
}
