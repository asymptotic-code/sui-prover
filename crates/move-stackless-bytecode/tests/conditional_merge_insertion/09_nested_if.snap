---
source: crates/move-stackless-bytecode/tests/testsuite.rs
---
============ initial translation from Move ================

[variant baseline]
public fun test::f($t0|cond1: bool, $t1|cond2: bool): u64 {
     var $t2|x#1#0: u64
     var $t3|y#1#0: u64
     var $t4: u64
     var $t5: u64
     var $t6: bool
     var $t7: u64
     var $t8: bool
     var $t9: u64
     var $t10: u64
     var $t11: u64
     var $t12: u64
  0: $t4 := 10
  1: $t2 := $t4
  2: $t5 := 20
  3: $t3 := $t5
  4: $t6 := move($t0)
  5: if ($t6) goto 6 else goto 15
  6: label L1
  7: $t7 := 99
  8: $t2 := $t7
  9: $t8 := move($t1)
 10: if ($t8) goto 11 else goto 15
 11: label L2
 12: $t9 := 100
 13: $t3 := $t9
 14: goto 15
 15: label L0
 16: $t10 := move($t2)
 17: $t11 := move($t3)
 18: $t12 := *($t10, $t11)
 19: return $t12
}

============ after pipeline `conditional_merge_insertion` ================

[variant baseline]
public fun test::f($t0|cond1: bool, $t1|cond2: bool): u64 {
     var $t2|x#1#0: u64
     var $t3|y#1#0: u64
     var $t4: u64
     var $t5: u64
     var $t6: u64
     var $t7: u64
     var $t8: u64
     var $t9: u64
     var $t10: u64
     var $t11: u64
     var $t12: u64
  0: $t4 := 10
  1: $t2 := $t4
  2: $t5 := 20
  3: $t3 := $t5
  4: if ($t0) goto 5 else goto 12
  5: label L1
  6: $t6 := 99
  7: $t9 := $t6
  8: if ($t1) goto 9 else goto 12
  9: label L2
 10: $t7 := 100
 11: $t10 := $t7
 12: label L0
     # conditional_merge_insertion: t11 := if_then_else(t0, t9, t2)
 13: $t11 := if_then_else($t0, $t9, $t2)
     # conditional_merge_insertion: merge assign t2 := t11
 14: $t2 := $t11
     # conditional_merge_insertion: t12 := if_then_else(t1, t10, t3)
 15: $t12 := if_then_else($t1, $t10, $t3)
     # conditional_merge_insertion: merge assign t3 := t12
 16: $t3 := $t12
 17: $t8 := *($t2, $t3)
 18: return $t8
}
